sdk/keyvault/keyvault-keys/src/index.ts

```diff
@@ -148,6 +147,86 @@ export {
   logger
 };
 
+/**
+ * Formats the received options into RequestOptions
+ */
+function toRequestOptionsBase(options?: CreateKeyOptions) {
+  let expandedOptions: RequestOptionsBase = {};
+
+  if (options) {
+    const requestOptions = operationOptionsToRequestOptionsBase(options);
+    const {
+      enabled,
+      notBefore,
+      expiresOn: expires,
+      hardwareProtected: hsm,
+      ...remainingOptions
+    } = requestOptions;
+    expandedOptions = {
+      ...remainingOptions,
+      keyAttributes: {
+        enabled,
+        notBefore,
+        expires,
+        hsm
+      }
+    };
+  }
+
+  return expandedOptions;
+}
+
+/**
+ * Prepares the request options and generates a tracer span.
+ * @param name Name of the tracer span.
+ * @param reqOptions Request options.
+ */
+function prepareTrace(name: string, reqOptions: RequestOptionsBase) {
+  const tracer = getTracer();
+  const span = tracer.startSpan(name, reqOptions && reqOptions.spanOptions);
+  span.setAttribute("az.namespace", "Microsoft.KeyVault");
+
+  let options = reqOptions;
+
+  // Setting the parent's span
+  if (span.isRecording()) {
+    const spanOptions = reqOptions.spanOptions || {};
+    options = {
+      ...reqOptions,
+      spanOptions: {
+        ...spanOptions,
+        parent: span.context(),
+        attributes: {
+          ...spanOptions.attributes,
+          "az.namespace": "Microsoft.KeyVault"
+        }
+      }
+    };
+  }
+
+  return { span, options };
+}
+
+/**
+ * Traces the given method, and returns the method's output without leaving active tracer spans.
+ * @param name Name of the tracer span.
+ * @param reqOptions Request options.
+ * @param method Method to wrap and trace.
+ */
+async function trace<T>(
```

Comment:
This is very similar to an approach Richard took in app-config.

The major downside is that if you are stepping through the code, you can't "step over" the tracing stuff, you instead have to step into this trace method to then step again back out to the method in the client, or set a BP inside the method itself and continue. This is made worse if you are debugging in generated async code for downlevel.

Reply:
What is the alternative though? All I can think of is ... decorators.